roomserver/internal/input/input_missing.go

```diff
@@ -83,16 +84,142 @@ func (t *missingStateReq) processEventWithMissingState(
+	// either not closed the gap, or we did but we still don't seem to
+	// know the events before the new event. Start by looking up the
+	// state at the event at the back of the gap and we'll try to roll
+	// forward the state first.
 	backwardsExtremity := newEvents[0]
 	newEvents = newEvents[1:]
 
+	resolvedState, err := t.lookupResolvedStateBeforeEvent(ctx, backwardsExtremity, roomVersion)
+	if err != nil {
+		return nil, fmt.Errorf("t.lookupState (backwards extremity): %w", err)
+	}
+
+	hadEvents := map[string]bool{}
+	t.hadEventsMutex.Lock()
+	for k, v := range t.hadEvents {
+		hadEvents[k] = v
+	}
+	t.hadEventsMutex.Unlock()
+
+	sendOutliers := func(resolvedState *gomatrixserverlib.RespState) error {
+		var outliers []*gomatrixserverlib.Event
+		outliers, err = resolvedState.Events()
+		if err != nil {
+			return fmt.Errorf("resolvedState.Events: %w", err)
+		}
+		var outlierRoomEvents []api.InputRoomEvent
+		for _, outlier := range outliers {
+			if hadEvents[outlier.EventID()] {
+				continue
+			}
+			outlierRoomEvents = append(outlierRoomEvents, api.InputRoomEvent{
+				Kind:   api.KindOutlier,
+				Event:  outlier.Headered(roomVersion),
+				Origin: t.origin,
+			})
+		}
+		for _, ire := range outlierRoomEvents {
+			_, err = t.inputer.processRoomEvent(ctx, t.db, &ire)
+			if err != nil {
+				if _, ok := err.(types.RejectedError); !ok {
+					return fmt.Errorf("t.inputer.processRoomEvent (outlier): %w", err)
+				}
+			}
+		}
+		return nil
+	}
+
+	// Send outliers first so we can send the state along with the new backwards
+	// extremity without any missing auth events.
+	if err = sendOutliers(resolvedState); err != nil {
+		return nil, fmt.Errorf("sendOutliers: %w", err)
+	}
+
+	// Now send the backward extremity into the roomserver with the
+	// newly resolved state. This marks the "oldest" point in the backfill and
+	// sets the baseline state for any new events after this.
+	stateIDs := make([]string, 0, len(resolvedState.StateEvents))
+	for _, event := range resolvedState.StateEvents {
+		stateIDs = append(stateIDs, event.EventID())
+	}
+
+	_, err = t.inputer.processRoomEvent(ctx, t.db, &api.InputRoomEvent{
+		Kind:          api.KindOld,
+		Event:         backwardsExtremity.Headered(roomVersion),
+		Origin:        t.origin,
+		HasState:      true,
+		StateEventIDs: stateIDs,
+		SendAsServer:  api.DoNotSendToOtherServers,
+	})
+	if err != nil {
+		if _, ok := err.(types.RejectedError); !ok {
+			return nil, fmt.Errorf("t.inputer.processRoomEvent (backward extremity): %w", err)
+		}
+	}
+
+	// Then send all of the newer backfilled events, of which will all be newer
+	// than the backward extremity, into the roomserver without state. This way
+	// they will automatically fast-forward based on the room state at the
+	// extremity in the last step.
+	for _, newEvent := range newEvents {
+		_, err = t.inputer.processRoomEvent(ctx, t.db, &api.InputRoomEvent{
+			Kind:         api.KindOld,
```

Comment:
Thinking of names for these... can we maybe do:
 - `KindOld` -> `KindTimeline`
 - `KindNew` -> `KindFwdsExtremity`?

Reply:
Suggest we do this in a separate PR — let's raise an issue and I can add it to my todo list.